volumina/tiling/cache.py

```diff
@@ -0,0 +1,299 @@
+            raise ValueError("size should non negative integer")
+
+    def set_size(self, value):
+        if self._size == value:
+            return
+
+        self._validate_size(value)
+
+        self._size = value
+
+        for sub in self._subsribers:
+            sub()
+
+    def subscribe(self, fn):
+        self._subsribers.append(fn)
+
+    def unsubsribe(self, fn):
+        try:
+            self._subsribers.remove(fn)
+        except ValueError:
+            pass
+
+
+class MultiCacheException(Exception):
+    pass
+
+
+class MultiCache:
+    """
+    A utility class for caching items in a dict-of-dicts
+    """
+
+    def __init__(self, first_uid, default_factory=lambda: None, policy=None):
+        self._policy = policy
+        self._policy.subscribe(self._clean)
+        self._caches = OrderedDict()
+        self._default_factory = default_factory
+        self.add(first_uid)
+
+    def add(self, uid) -> None:
+        if uid not in self._caches:
+            cache = defaultdict(self._default_factory)
+            self._caches[uid] = cache
+        else:
+            raise MultiCacheException("MultiCache.add: uid %s is already in use" % str(uid))
+
+        # remove oldest cache, if necessary
+        self._clean()
+
+    def __contains__(self, key):
+        return key in self._caches
+
+    def __getitem__(self, key):
+        return self._caches[key]
+
+    def __iter__(self):
+        return iter(self._caches)
+
+    def __len__(self):
+        return len(self._caches)
+
+    def touch(self, uid):
+        self._caches.move_to_end(uid)
+
+    @property
+    def maxsize(self):
+        return self._policy.size
+
+    def _evict_one(self):
+        self._caches.popitem(last=False)  # removes item in FIFO order
+
+    def _clean(self):
+        while len(self._caches) > self.maxsize:
+            self._evict_one()
+
+
+class TilesCache:
+    """
+    Contains the following caches, with convenience accessor functions for each.
+
+        layerCache: A cache of 'layers', i.e. for every patch a QImage or QGraphicsItem
+                    for every "image source" in the stack
+
+        tileCache: A cache of 'tiles', i.e. the blended QImage objects
+                   that were created by combining all QImage layers from layerCache for a given patch.
+                   (The QGraphicsItem layers do not contribute to the composite tiles in the tileCache.
+                   They are merely stored.)
+
+        layerCacheDirty: A cache of dirty bits for all layers in layerCache
+        layerCacheTimestamp: A cache of timestamps to track how recently each layer was needed.
+
+        tileCacheDirty: A cache of dirty bits for the composite tiles
+                        (i.e. for a given patch, if a single layer in the patch
+                        is dirty, then the tile for that patch is dirty)
+    """
+
+    def __init__(self, first_stack_id, sims, maxstacks):
+        self._lock = threading.Lock()
```

Comment:
This probably will not make a difference, but I think it's safe to use a reentrant lock here.
```suggestion
        self._lock = threading.RLock()
```

Reply:
This change crashes volumina tests :(